File: file/src/main/scala/farjs/file/popups/FileViewHistoryPopup.scala
package farjs.file.popups

import farjs.file.FileViewHistory
import farjs.file.FileViewHistory.fileViewsHistoryKind
import farjs.filelist.history.HistoryProvider
import farjs.ui.popup._
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

object FileViewHistoryPopup extends FunctionComponent[FileViewHistoryPopupProps] {

  private[popups] var listPopup: ReactClass = ListPopup
  
  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider()
    val (maybeItems, setItems) = useState(Option.empty[js.Array[FileViewHistory]])
    val props = compProps.plain

    useLayoutEffect({ () =>
      for {
        fileViewsHistory <- historyProvider.get(fileViewsHistoryKind).toFuture
        items <- fileViewsHistory.getAll().toFuture
      } yield {
        setItems(Some(items.flatMap(h => FileViewHistory.fromHistory(h).toOption)))
      }
      ()
    }, Nil)

    maybeItems.map { items =>
      <(listPopup)(^.plain := ListPopupProps(
        title = "File view history",
        items = items.map { item =>
          val prefix =
            if (item.params.isEdit) "Edit: "
            else "View: "
          s"$prefix${item.path}"
        },
        onAction = { index =>
          props.onAction(items(index))
        },
        onClose = props.onClose,
        selected = math.max(items.length - 1, 0),
        itemWrapPrefixLen = 9
      ))()
    }.orNull
  }
}

File: file/popups/FileViewHistoryPopup.mjs
/**
 * @typedef {import("../FileViewHistory.mjs").FileViewHistory} FileViewHistory
 */
import React, { useLayoutEffect, useState } from "react";
import ListPopup from "@farjs/ui/popup/ListPopup.mjs";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";
import FileViewHistory from "../FileViewHistory.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  onAction(h: FileViewHistory): void;
 *  onClose(): void;
 * }} FileViewHistoryPopupProps
 */

/**
 * @param {FileViewHistoryPopupProps} props
 */
const FileViewHistoryPopup = (props) => {
  const { listPopup } = FileViewHistoryPopup;

  const historyProvider = HistoryProvider.useHistoryProvider();
  const [maybeItems, setItems] = useState(
    /** @type {readonly FileViewHistory[] | undefined} */ (undefined)
  );

  const readHistory = async () => {
    const fileViewsHistory = await historyProvider.get(
      FileViewHistory.fileViewsHistoryKind
    );
    const items = await fileViewsHistory.getAll();
    const historyItems = items.reduce((res, item) => {
      const h = FileViewHistory.fromHistory(item);
      if (h) {
        res.push(h);
      }
      return res;
    }, /** @type {FileViewHistory[]} */ ([]));
    setItems(historyItems);
  };

  useLayoutEffect(() => {
    readHistory();
  }, []);

  if (maybeItems !== undefined) {
    return h(listPopup, {
      title: "File view history",
      items: maybeItems.map((item) => {
        const prefix = item.params.isEdit ? "Edit: " : "View: ";
        return `${prefix}${item.path}`;
      }),
      onAction: (index) => {
        props.onAction(maybeItems[index]);
      },
      onClose: props.onClose,
      selected: Math.max(maybeItems.length - 1, 0),
      itemWrapPrefixLen: 9,
    });
  }

  return null;
};

FileViewHistoryPopup.displayName = "FileViewHistoryPopup";
FileViewHistoryPopup.listPopup = ListPopup;

export default FileViewHistoryPopup;
